tcpping/tcpping.go

```diff
@@ -0,0 +1,95 @@
+package tcpping
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	errors "github.com/pkg/errors"
+	tcp "github.com/tevino/tcp-shaker"
+)
+
+type Result struct {
+	Address string
+	Latency int
+	PingErr error
+}
+
+func (cr *Result) String() string {
+	return fmt.Sprintf(
+		"Result{address: %s, latency: %d, error: %s}\n",
+		cr.Address, cr.Latency, cr.PingErr)
+}
+
+func runCheck(c *tcp.Checker, address string, resChan chan Result, timeout time.Duration) {
+	// mesaure RTT
+	start := time.Now()
+	// TCP Ping
+	err := c.CheckAddr(address, timeout)
+	// measure RTT
+	elapsed := time.Since(start)
+	latency := int(elapsed.Milliseconds())
+
+	if err != nil { // don't confuse users with valid latency values on error
+		latency = -1
+		if err == tcp.ErrTimeout {
+			err = errors.Wrap(err, "tcp check timeout")
+		} else {
+			switch err.(type) {
+			case tcp.ErrConnect:
+				err = errors.Wrap(err, "unable to connect")
+			default:
+				err = errors.Wrap(err, "unknown error")
+			}
+		}
+	}
+
+	// return results
+	resChan <- Result{
+		Address: address,
+		Latency: latency,
+		PingErr: err,
+	}
+}
+
+func waitForResults(resChan chan Result, expected int, timeout time.Duration) ([]Result, error) {
+	var checks = []Result{}
+	for {
+		select {
+		case result := <-resChan:
+			checks = append(checks, result)
+			if len(checks) == expected {
+				return checks, nil
+			}
+		case <-time.After(timeout):
+			return checks, errors.New("results wait timeout")
+		}
+	}
+}
+
+func CheckHosts(addresses []string, timeout time.Duration) ([]Result, error) {
+	c := tcp.NewChecker()
+
+	// stop the checking loop when function exists
+	ctx, stopChecker := context.WithCancel(context.Background())
+	defer stopChecker()
+
+	// loop that queries Epoll and pipes events
+	go func() {
+		// TODO handle error from this?
+		c.CheckingLoop(ctx)
```

Comment:
```suggestion
		errCh <- c.CheckingLoop(ctx)
```

where `errCh := make(chan error, 1)`

Reply:
But where should I read from that `errCh`? In `waitForResults()`? After it finishes?

Reply:
Ok, found a solution:
```go
func waitForResults(errCh <-chan error, resCh <-chan Result) (results []Result, err error) {
	for {
		select {
		case err = <-errCh:
			return nil, err
		case res, ok := <-resCh:
			if !ok {
				return
			}
			results = append(results, res)
		}
	}
}
```